planner/core/rule_column_pruning.go

```diff
@@ -137,9 +137,15 @@ func (la *LogicalAggregation) PruneColumns(parentUsedCols []*expression.Column)
 
 func pruneByItems(old []*util.ByItems) (new []*util.ByItems, parentUsedCols []*expression.Column) {
 	new = make([]*util.ByItems, 0, len(old))
+	seen := make(map[string]expression.Expression, len(old))
 	for _, byItem := range old {
+		hash := string(byItem.Expr.HashCode(nil))
+		possibleDupe, hashMatch := seen[hash]
+		seen[hash] = byItem.Expr
 		cols := expression.ExtractColumns(byItem.Expr)
-		if len(cols) == 0 {
+		if hashMatch && possibleDupe.Equal(nil, byItem.Expr) {
```

Comment:
It can be simplified to the following:

```suggestion
		if hashMatch {
```